[Feature][SQL Transform] Check SQL Cast will fail or not during the SQL parsing stage

liuwei178 opened the pull request:
- Removed unnecessary type conversion logic in SystemFunction
- Added check on data type cast in ZetaSQLType.

<!--

Thank you for contributing to SeaTunnel! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

## Contribution Checklist
  - Make sure that the pull request corresponds to a [GITHUB issue](https://github.com/apache/seatunnel/issues).
  - Name the pull request in the form "[Feature] [component] Title of the pull request", where *Feature* can be replaced by `Hotfix`, `Bug`, etc.
  - Minor fixes should be named following this pattern: `[hotfix] [docs] Fix typo in README.md doc`.
-->

### Purpose of this pull request

<!-- Describe the purpose of this pull request. For example: This pull request adds checkstyle plugin.-->


### Does this PR introduce _any_ user-facing change?

<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released SeaTunnel versions or within the unreleased branches such as dev.
If no, write 'No'.
If you are adding/modifying connector documents, please follow our new specifications: https://github.com/apache/seatunnel/issues/4544.
-->


### How was this patch tested?

<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If you are adding E2E test cases, maybe refer to https://github.com/apache/seatunnel/blob/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-cdc-mysql-e2e/src/test/resources/mysqlcdc_to_mysql.conf, here is a good example.
-->


### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  2. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
  3. Add ci label in [label-scope-conf](https://github.com/apache/seatunnel/blob/dev/.github/workflows/labeler/label-scope-conf.yml)
  4. Add e2e testcase in [seatunnel-e2e](https://github.com/apache/seatunnel/tree/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/)
  5. Update connector [plugin_config](https://github.com/apache/seatunnel/blob/dev/config/plugin_config)

Commit:
[Feature][SQL Transform]: 修复 cast 函数的类型转换逻辑
- 移除了 SystemFunction 中不必要的类型转换逻辑
- 在 ZetaSQLType 中增加了对不同数据类型转换的限制

Hisoka-X commented:
Thanks @liuwei178 . Please follow the guide to open CI on your fork repository. https://github.com/apache/seatunnel/pull/9600/checks?check_run_id=46369080578

Hisoka-X reviewed (commented):
Please add test case

Commit:
Trigger CI workflow

Commit:
[Feature][SQL Transform]: code format

Commit:
[Feature][SQL Transform]: code format

Commit:
[Feature][SQL Transform]:

liuwei178 committed:
Merge branch 'apache:dev' into dev

Commit:
Merge remote-tracking branch 'origin/dev' into dev

Commit:
[Feature][SQL Transform]: add test case

liuwei178 commented:
i added the test and enabled the CI. Please take some time to review them when you're available @Hisoka-X

liuwei178 commented:
<img width="1802" height="457" alt="image" src="https://github.com/user-attachments/assets/2e5afc41-1d0f-4b0e-b8ed-f8d967a7db88" />
<img width="1604" height="653" alt="image" src="https://github.com/user-attachments/assets/c8f97598-3090-496f-ae54-262d40a9b7ac" />
What should I do if some unmodified modules report a timeout during the CI phase?

Hisoka-X commented:
> What should I do if some unmodified modules report a timeout during the CI phase?

Please retrigger the failed ci. Thanks

Hisoka-X commented on seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/sql/zeta/ZetaSQLType.java:
```diff
@@ -85,6 +86,24 @@ public class ZetaSQLType {
     public static final String TIME = "TIME";
     public static final String BOOLEAN = "BOOLEAN";
 
+    public static final List<SqlType> DATETIME_CAST_TYPES =
```
Thanks @liuwei178 . Can we refactor it and move the cast-related logic and validation rules into separate classes?

liuwei178 replied:
Okay, I'll refactor it as soon as possible

liuwei178 replied:
@Hisoka-X   hi，Please help check which parts still need to adjust

liuwei178 committed:
Merge branch 'apache:dev' into dev

liuwei178 committed:
Merge branch 'apache:dev' into dev

Commit:
[Feature][SQL Transform]:
-Added CastFunction class to handle type conversion operations
-Optimized type conversion logic to support more data type conversions

Commit:
[Feature][SQL Transform]: -bigint to int Illegal

Commit:
[Feature][SQL Transform]:
-add long type

Hisoka-X reviewed (approved):
LGTM. Thanks @liuwei178

Carl-Zhou-CN commented on seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/sql/zeta/functions/CastFunction.java:
```diff
@@ -0,0 +1,161 @@
+                return BasicType.STRING_TYPE;
+            case BYTE:
+            case TINYINT:
+                if (SqlType.TINYINT.equals(originType) || SqlType.STRING.equals(originType)) {
```
For validation types, are they referring to the types that we currently have the capability to convert?

liuwei178 replied:
Yes, these are just some pre-validation steps added to our existing capabilities, for example, when converting Timestamp to SMALLINT is illegal.    @Carl-Zhou-CN 
Link to https://github.com/apache/seatunnel/issues/9538

Carl-Zhou-CN replied:
Ok. Thank you for your answer

Carl-Zhou-CN reviewed (approved):
+1
